lib/Parse/ParseType.cpp

```diff
@@ -545,38 +545,54 @@ bool Parser::parseGenericArguments(SmallVectorImpl<TypeRepr*> &Args,
 ///   type-identifier:
 ///     identifier generic-args? ('.' identifier generic-args?)*
 ///
-ParserResult<TypeRepr> Parser::parseTypeIdentifier() {
-  if (Tok.isNot(tok::identifier) && Tok.isNot(tok::kw_Self)) {
-    // is this the 'Any' type
-    if (Tok.is(tok::kw_Any)) {
-      return parseAnyType();
-    } else if (Tok.is(tok::code_complete)) {
-      if (CodeCompletion)
-        CodeCompletion->completeTypeSimpleBeginning();
-      // Eat the code completion token because we handled it.
-      consumeToken(tok::code_complete);
-      return makeParserCodeCompletionResult<IdentTypeRepr>();
-    }
+/// SWIFT_ENABLE_TENSORFLOW: Added `isParsingQualifiedDeclName` flag.
+/// The `isParsingQualifiedDeclName` flag controls whether parsing is done as if
+/// this type identifier is the prefix to a qualified declaration name. If true,
```

Comment:
can you explain what "qualified declaration name" is, and/or give an example?

also, why do we need to add this knob -- what's special about the AD expr type compared to other expr types?

(the change to this func seems pretty invasive and is more likely to cause merge conflicts than other s4tf changes. so it'd be useful to spend extra care and assess if there are better alternatives.)

Reply:
Thanks for the pointers! I did read the PR description. I'll try and read other pre-existing PR comments in future reviews as well (somtimes reviewers may not be reading all review exchanges with other parties).

Name overloading is not special to AD. Usually a language has precedence rules to help decide which one to choose, like what's being implemented here -- essentially when there are multiple matches, pick the one with the scope closest to the use-site. 

I'm guess there is an pre-existing impl in swift (is it in type checker, for ranking candidates and picking the best fit one?). it'd be great to try and find that and reuse. please file a bug to track this.

Reply:
I mentioned overloaded names as the reason for parsing a qualified name rather than an unqualified one.
However, the actual overload resolution isn't handled here - that's done in CSGen via an existing function [`TypeChecker::lookupFuncDecl`](https://github.com/apple/swift/blob/4316cf9f25cb441ef0108a703874dc85058c60e1/lib/Sema/CSGen.cpp#L1468).
Here, the base type and function name are parsed. These are passed to `lookupFuncDecl` as arguments.

Reply:
Thanks for weighing in and asking questions! I think the examples you listed ("name overloading", "precedence rules", etc) are largely irrelevant to the topic being discussed here. This particular code piece is adding backtracking logic to enable code sharing and parsing a fully specified decl name, not related to AD or overload resolution. "Qualified decl name" has a well-defined meaning throughout Swift's AST, and represents a complete lookup path. And there is no pre-existing parser logic to parse a qualified decl name **without** parsing an expression.

In this particular case, the code author has implementation experience and established knowledge about how parser and type checking works, and the implementation a) does generally follow the established convention and b) does have clear explanation of each decision made. Before drawing speculations based largely on assumptions, I would appreciate it if you read more about the how Parse and Sema work.

Reply:
From a follow-up discussion with Dan, I captured these points:

- Before this change, there is no concept/support of a qualified flavor of a decl name in Swift compiler. A func name like `A.B` would be an expression (DeclRefExpr), and name overloading is solved with that mechanism.
- But for AD infra, using an expression for the adjoint func name is evaluated and decided against. Hence we are extending decl name to support the qualified flavor.

Thanks for the additional context and clarification!

Reply:
Great! We'll just have to take extra care to incorporate this upstream.